### ISSUE TITLE
Operators should take advantage of sorted chunks/mark chunks as sorted

### ISSUE THREAD

**Issue by cmfcmf** [hyrise/hyrise#1519]
After #1394 is merged, Chunks can now indicate if they have a sorted segment. Currently, only the ColumnVsValueScan takes advantage of this flag. This issue is here as a reminder 
1. to take advantage of this flag in other operators
2. to set the flag where a chunk is actively sorted
3. to set the flag where a chunk of ReferenceSegments pointing to an already sorted chunk is created and the relative order remains the same

Regarding 1), here's an incomplete list of operators that might benefit from the flag:
- ColumnIsNull scan
- ColumnVsColumn scan
- Aggregate
- SortMergeJoin
- Sort

Regarding 2), here's a list of places where a chunk could be marked as sorted:
- Sort operator
- SortMergeJoin
- During chunk encoding, a chunk could either be sorted by a single segment (there would need to be statistics/heuristics to figure out which segment to sort) or at least all segments could be tested to see whether they are already sorted (most likely this will only work for things like auto incrementing IDs)

Regarding 3), here's an incomplete list of operators that leave the relative order in tact and should therefore set the sort flag on the output chunks if the input chunks were sorted:
- Limit
- Validate
- Projection
- TableScan
- Alias

**Comment by mrzzzrm** [comment]
Setting and especially forwarding flags on Tables/Chunks/Segments is not something we should do in Hyrise. It requires Operators, which *should* be single-concern algorithms, to think about what everybody else might do. Think what would happen if we had 4, 5, even more flags that all operators would need to consider, could maybe only forward in specific situations and so forth. It would be a lot of logic in ALL operators for doing things they are not concerned with.

We had this for "is_validated". It was bad. We have it for the PosList's "references_only_a_single_chunk" guarantee, I trip over that regularly, though I wouldn't have a better idea for it...

Anyway, long story short, we need a different solution here. My idea would be an optimizer pass that knows about the behaviour of all operators and propagates/removes sortedness and other properties. I have heard this concept being called "InterestingProperties". Awkward name, but if we stick with it, the `AbstractLQPNode` and `AbstractOperator` could get a `InterestingProperties _output_properties` field.

**Comment by Bouncner** [comment]
I see problems with both approaches. For example, the sort-merge join can give certain guarantees for a particular configuration but not for others (this could be requested by the optimizer though).
Another fact are decision made at runtime. For join-heavy workloads like the JOB,  things as the proxyjoin could be beneficial, but how would you transport that knowledge?

**Comment by mrzzzrm** [comment]
Admittedly, with Proxy Operators the Optimizer-sets-properties approach falls apart.

**Comment by Bensk1** [comment]
I agree that we might need a more holistic solution for that problem and interesting properties in general. But I don't know whether not adding any information to tables or chunks at all is a solution.

Sortedness is a fundamental concept and could not only occur during query processing. We might need a way of keeping that information across query execution boundaries. We discussed, for example, the opportunity for the database system to decide that a table/chunk should be sorted in a certain way based on the whole database workload. This information can be used by the optimizer, but it needs to be stored with the data.

**Comment by mrzzzrm** [comment]
Okay, if forwarding Chunk properties in operators is unavoidable we have to *force* the operators to think about them. It should be impossible to construct a Chunk without specifying its properties, i.e., by extending its constructor with a non-optional parameter. 
If we allow constructing Chunks without setting the properties people will just miss it and we gain nothing (but confusion and chaos)

**Comment by Bouncner** [comment]
That’s a good point. Though I cannot think of other “interesting properties” right now, we might need to take some time and think about an `interestingpropertiesstruct` that covers multiple flags!?!

**Comment by mrzzzrm** [comment]
> we might need to take some time and think

yes!

**Comment by Bensk1** [comment]
> Okay, if forwarding Chunk properties in operators is unavoidable

I don't know about the forwarding and the operator part. I meant mostly that setting this information in a Table/Chunk is most certainly necessary.

**Comment by mrks** [comment]
Another thought: As of #2000, 
* `Table::_value_clustered_by` is not forwarded by operators like Projection.
* `JoinHash` does not set it, but should be able to set it just like `JoinSortMerge`: result_table->set_value_clustered_by({first_join_column, second_join_column});
* Doing so might make AggregateSort the operator of choice in [TPC-H Q10](https://hyrise-ci.epic-hpi.de/job/hyrise/job/hyrise/job/master/lastStableBuild/artifact/query_plans/tpch/TPC-H_10-PQP.svg)
